src/CustomElements/v1/CustomElements.js

```diff
@@ -0,0 +1,597 @@
+        this._whenDefinedMap.delete(localName);
+      }
+    },
+
+    /**
+     * Returns the constructor defined for `name`, or `null`.
+     *
+     * @param {string} name
+     * @return {Function|undefined}
+     */
+    get: function(name) {
+      // https://html.spec.whatwg.org/multipage/scripting.html#custom-elements-api
+      var def = this._definitions.get(name);
+      return def ? def.constructor : undefined;
+    },
+
+    /**
+     * Returns a `Promise` that resolves when a custom element for `name` has
+     * been defined.
+     *
+     * @param {string} name
+     * @return {Promise}
+     */
+    whenDefined: function(name) {
+      // https://html.spec.whatwg.org/multipage/scripting.html#dom-customelementsregistry-whendefined
+      if (!customNameValidation.test(name)) {
+        return Promise.reject(
+          new SyntaxError(`The element name '${name}' is not valid.`));
+      }
+      if (this._definitions.has(name)) {
+        return Promise.resolve();
+      }
+      var deferred = {
+        promise: null,
+      };
+      deferred.promise = new Promise(function(resolve, _) {
+       deferred.resolve = resolve;
+      });
+      this._whenDefinedMap.set(name, deferred);
+      return deferred.promise;
+    },
+
+    /**
+     * Causes all pending mutation records to be processed, and thus all
+     * customization, upgrades and custom element reactions to be called.
+     * `enableFlush` must be true for this to work. Only use during tests!
+     */
+    flush: function() {
+      if (this.enableFlush) {
+        console.warn("flush!!!");
+        this._observers.forEach(function(observer) {
+          this._handleMutations(observer.takeRecords());
+        }, this);
+      }
+    },
+
+    _setNewInstance: function(instance) {
+      this._newInstance = instance;
+    },
+
+    /**
+     * Observes a DOM root for mutations that trigger upgrades and reactions.
+     * @private
+     */
+    _observeRoot: function(root) {
+      root.__observer = new MutationObserver(this._handleMutations.bind(this));
+      root.__observer.observe(root, {childList: true, subtree: true});
+      if (this.enableFlush) {
+        // this is memory leak, only use in tests
+        this._observers.add(root.__observer);
+      }
+    },
+
+    /**
+     * @private
+     */
+    _unobserveRoot: function(root) {
+      if (root.__observer) {
+        root.__observer.disconnect();
+        root.__observer = null;
+        if (this.enableFlush) {
+          this._observers.delete(root.__observer);
+        }
+      }
+    },
+
+    /**
+     * @private
+     */
+    _handleMutations: function(mutations) {
+      for (var i = 0; i < mutations.length; i++) {
+        var mutation = mutations[i];
+        if (mutation.type === 'childList') {
+          // Note: we can't get an ordering between additions and removals, and
+          // so might diverge from spec reaction ordering
+          this._addNodes(mutation.addedNodes);
```

Comment:
As implemented, `_addNodes` always connects elements after customization, which is not correct for HTML imports, or additions of shadow root children of detached hosts.

You can bind `root` into `_handleMutations`, and then tell `_addNodes` to only connect if the root is `window.document`, or if `root.host` is currently attached.